src/storage.rs

```diff
@@ -96,166 +113,242 @@ pub trait Storage {
-    /// can be used to reconstruct the state at that point.
-    /// If any configuration changes have been made since the last compaction,
-    /// the result of the last apply_conf_change must be passed in.
-    pub fn create_snapshot(
-        &mut self,
-        idx: u64,
-        cs: Option<ConfState>,
-        pending_membership_change: Option<ConfChange>,
-        data: Vec<u8>,
-    ) -> Result<&Snapshot> {
-        if idx <= self.snapshot.get_metadata().get_index() {
-            return Err(Error::Store(StorageError::SnapshotOutOfDate));
-        }
-
-        let offset = self.entries[0].get_index();
-        if idx > self.inner_last_index() {
-            panic!(
-                "snapshot {} is out of bound lastindex({})",
-                idx,
-                self.inner_last_index()
-            )
-        }
-        self.snapshot.mut_metadata().set_index(idx);
-        self.snapshot
-            .mut_metadata()
-            .set_term(self.entries[(idx - offset) as usize].get_term());
-        if let Some(cs) = cs {
-            self.snapshot.mut_metadata().set_conf_state(cs)
-        }
-        if let Some(pending_change) = pending_membership_change {
-            let meta = self.snapshot.mut_metadata();
-            meta.set_pending_membership_change(pending_change.get_configuration().clone());
-            meta.set_pending_membership_change_index(pending_change.get_start_index());
+    fn inner_snapshot(&self) -> Snapshot {
+        let mut snapshot = Snapshot::new();
+
+        // Use the latest applied_idx to construct the snapshot.
+        let applied_idx = self.raft_state.hard_state.get_commit();
+        let term = self.raft_state.hard_state.get_term();
+        snapshot.mut_metadata().set_index(applied_idx);
+        snapshot.mut_metadata().set_term(term);
+
+        // Find the latest configuration state before applied index.
+        let i = (0..self.raft_state.conf_states.len())
+            .take_while(|i| self.raft_state.conf_states[*i].index <= applied_idx)
+            .last()
+            .unwrap();
+        if self.raft_state.conf_states[i].in_membership_change {
+            let cs = self.raft_state.conf_states[i - 1].clone();
+            snapshot.mut_metadata().set_conf_state(cs.conf_state);
+            let cs = self.raft_state.conf_states[i].clone();
+            snapshot
+                .mut_metadata()
+                .set_pending_membership_change(cs.conf_state);
+            snapshot
+                .mut_metadata()
+                .set_pending_membership_change_index(cs.index);
+        } else {
+            let cs = self.raft_state.conf_states[i].clone();
+            snapshot.mut_metadata().set_conf_state(cs.conf_state);
         }
-        self.snapshot.set_data(data);
-        Ok(&self.snapshot)
+        snapshot
     }
 
-    /// Discards all log entries prior to compact_index.
+    /// Discards all log entries prior to compact_index (included).
```

Comment:
what does the included mean? such as [0, 1, 2, 3],  it means `compact(3)` will remains [3] but allow to compact(4) or `compact(3)` will remains empty.

Reply:
Stale comment. I will remove it.